discord/app/cache.py

```diff
@@ -0,0 +1,306 @@
+        ...
+
+    # polls
+
+    async def get_all_polls(self) -> list[Poll]:
+        ...
+
+    async def get_poll(self, message_id: int) -> Poll:
+        ...
+
+    async def store_poll(self, poll: Poll, message_id: int) -> None:
+        ...
+
+    # private channels
+
+    async def get_private_channels(self) -> list[PrivateChannel]:
+        ...
+
+    async def get_private_channel(self, channel_id: int) -> PrivateChannel:
+        ...
+
+    async def store_private_channel(self, channel: PrivateChannel, channel_id: int) -> None:
+        ...
+
+    # dm channels
+
+    async def get_dm_channels(self) -> list[DMChannel]:
+        ...
+
+    async def get_dm_channel(self, channel_id: int) -> DMChannel:
+        ...
+
+    async def store_dm_channel(self, channel: DMChannelPayload, channel_id: int) -> DMChannel:
+        ...
+
+    def clear(self, views: bool = True) -> None:
+        ...
+
+class MemoryCache(Cache):
+    def __init__(self, max_messages: int | None = None, *, state: ConnectionState):
+        self._state = state
+        self.max_messages = max_messages
+        self.clear()
+
+    def clear(self, views: bool = True) -> None:
+        self._users: dict[int, User] = {}
+        self._guilds: dict[int, Guild] = {}
+        self._polls: dict[int, Poll] = {}
+        self._stickers: dict[int, list[GuildSticker]] = {}
+        if views:
+            self._views: dict[str, View] = {}
+        self._modals: dict[str, Modal] = {}
+        self._messages: Deque[Message] = deque(maxlen=self.max_messages)
+
+        self._emojis = dict[str, GuildEmoji | AppEmoji] = {}
+
+        self._private_channels: OrderedDict[int, PrivateChannel] = OrderedDict()
+        self._private_channels_by_user: dict[int, DMChannel] = {}
+
+    # users
+    async def get_all_users(self) -> list[User]:
+        return list(self._users.values())
+
+    async def store_user(self, payload: UserPayload) -> User:
+        user_id = int(payload["id"])
+        try:
+            return self._users[user_id]
+        except KeyError:
+            user = User(state=self, data=payload)
+            if user.discriminator != "0000":
+                self._users[user_id] = user
+                user._stored = True
+            return user
+
+    async def delete_user(self, user_id: int) -> None:
+        self._users.pop(user_id, None)
+
+    async def get_user(self, user_id: int) -> User:
+        return self._users.get(user_id)
+
+    # stickers
+
+    async def get_all_stickers(self) -> list[GuildSticker]:
+        return list(self._stickers.values())
+
+    async def get_sticker(self, sticker_id: int) -> GuildSticker:
+        return self._stickers.get(sticker_id)
+
+    async def store_sticker(self, guild: Guild, data: GuildStickerPayload) -> GuildSticker:
+        sticker = GuildSticker(state=self._state, data=data)
+        try:
+            self._stickers[guild.id].append(sticker)
+        except KeyError:
+            self._stickers[guild.id] = sticker
```

Comment:
Shouldn't this be `self._stickers[guild.id] = [sticker]` ? Also maybe could use defaultdict